### PR TITLE
Popup 238 threejs interface based spatialdraw tool

### PR DESCRIPTION
**Pull request by ptc-rdeleeuw**
refactoring drawing tool to make use of the three.js interface

should be merged together with:
https://github.com/ptcrealitylab/vuforia-spatial-toolbox-userinterface/pull/475
https://github.com/ptcrealitylab/vuforia-spatial-edge-server/pull/821
https://github.com/ptcrealitylab/vuforia-spatial-remote-operator-addon/pull/95

### CODE REVIEW

File: tools/simpleModel/SimpleModelWorker.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        if (typeof message !== 'object') {
+            return;
+        }
+        if (message.hasOwnProperty('name')) {
+            if ((message.name === 'groundPlaneCallback') && (this.groundPlaneContainerObj !== null)) {
+                setMatrixFromArray(this.groundPlaneContainerObj.matrix, message.modelViewMatrix);
+                if (!this.isGroundPlaneFound) {
+                    this.isGroundPlaneFound = true;
+                }
+            } else if ((message.name === 'modelViewCallback') ) {
```
Comment (hobinjk-ptc) [inline tools/simpleModel/SimpleModelWorker.js:44]: nit: extra parentheses and an extra space

**Comment by hobinjk-ptc** [comment]
After deleting a spatialDraw tool that never fully initialized (unsure of what happened with it, I think it was from an earlier round of testing this?), I now can successfully use the spatialDraw tool. I do still see the `glRenderer.js:453 Server is in a wrong state to ask for more frames. serverState: 6` error but it isn't followed by the GL_INVALID_OPERATION

**Comment by hobinjk-ptc** [comment]
Nevermind, I added a new spatialDraw tool and that one is experiencing the error and only intermittently draws lines

**Comment by hobinjk-ptc** [comment]
Sending a video on Teams for better information than my words :D

**Comment by ptc-rdeleeuw** [comment]
I've experienced the same behaviour and thought I solved it. Apparently there are still cases where the system deadlocks, server state 6 is STATE_FRAME_SYNC, which means that the server tries to ask for a new frame while the worker thread is in a locked state and is waiting for a reply. I suspect this to happen when a frame is canceled based on it's duration, but the worker thread was waiting for a response. Since render command buffers cannot have these kind of commands since they have to be repeatable without workerthread communication, the canceled buffer must be a resource buffer. I think we might need to make sure that command buffers are atomic, either they get executed or they don't. 

I have found that the duplicate context strategy is more robust since that is what we are using on the iPhone, we can safely disable the threaded strategy by letting useThreads (glCommandBuffer@ln 1889) return false. We can merge and open a new issue to look deeper into this deadlock and explorer solutions.

**Comment by hobinjk-ptc** [comment]
Okay, sounds reasonable. I'm good with moving forwards with merging

**Review by hobinjk-ptc — APPROVED** [review]

**Comment by ptc-rdeleeuw** [comment]
I will disable the threaded approach and create a new issue for solving the deadlock. this will make it "stable". I use quotation marks here because both contexts should never be lost/restored. the helper context for compiling shaders, cannot currently restore itself properly. If it could we would have another problem, let's look at the cases:
main context restored, helper context restored: everything works hooray!
main context restored, helper context destroyed: we cannot perform actions on shaders because there is no helper context
main context destroyed, helper context restored: to prevent resource leaks, we should act as if the helper context is destroyed
main context destroyed, helper context destroyed: everything works hooray!
because javascript is single threaded for this strategy and we cannot influence the order in which contexts are restored/destroyed by the browser/driver we can never guarantee that the bad state of the system occurs. 
There is no goo way to mitigate this because a context can get destroyed without warning, if a context gets lost, all gl calls will return an error and in order to recover we need to run code. This code should not be interrupted, but that's where the time limit for each tool comes in again, which might interrupt the execution of this recovery code.

Since the previous implementation deemed stable has the same problem, I'm fine with merging. Knowing that the real solution is to go to a higher abstraction level in order to make a proper 3d engine handle context loss and shader compilation.

**Pull request merged**